# 1. Viết class và cài phương thức softmax.
Trong pytorch, torch.nn.Module là lớp cơ bản để từ đó xây dựng lên các mô hình hoặc các phương thức kích hoạt (activation funtion) như sigmoid, softmax,... Trong phần này, chúng ta xây dựng class Softmax và softmax_stable nhận đầu vào là mảng 1 chiều (tensor 1 chiều) dựa vào kế thừa từ lớp Module và ghi đè vào phương thức forward() theo công thức sau đây:

$ softmax(x_j) = \frac{exp(x_j)}{\sum_{j=1}^{n}exp(x_j)}$


$ softmax\_stable(x_j) = \frac{exp(x_j-c)}{\sum_{j=1}^{n}exp(x_j-c)}$

with $c = max(x)$
    




In [6]:
from zutils import Softmax, Softmax_stable
import torch

In [7]:
# Test softmax

data = torch.Tensor ([1, 2, 3])
softmax = Softmax()
output = softmax(data)
output

tensor([0.0900, 0.2447, 0.6652])

In [8]:
# Test softmax_stable
data = torch.Tensor ([1, 2, 3])
softmax_stable = Softmax_stable()
output = softmax_stable(data)
output

tensor([0.0900, 0.2447, 0.6652])

# 2. Một Ward (phường) gồm có name (string) và danh sách của mọi người trong Ward.
Một người person có thể là student, doctor, hoặc teacher. Một student gồm có name,
yob (int) (năm sinh), và grade (string). Một teacher gồm có name, yob, và subject
(string). Một doctor gồm có name, yob, và specialist (string). Lưu ý cần sử dụng a
list để chứa danh sách của mọi người trong Ward.

1. Cài đặt các class Student, Doctor, và Teacher theo mô tả trên. Thực hiện phương thức
describe() method để in ra tất cả thông tin của các object.

2. Viết add_person(person) method trong Ward class để add thêm một người mới với nghề
nghiệp bất kỳ (student, teacher, doctor) vào danh sách người của ward. Tạo ra một ward
object, và thêm vào 1 student, 2 teacher, và 2 doctor. Thực hiện describe() method để in ra tên ward (name) và toàn bộ thông tin của từng người trong ward.

3. Viết count_doctor() method để đếm số lượng doctor trong ward.

4. Viết sort_age() method để sort mọi người trong ward theo tuổi của họ với thứ tự tăng dần. 

    (hint: Có thể sử dụng sort của list hoặc viết thêm function đều được)

5. Viết compute_average() method để tính trung bình năm sinh của các teachers trong ward.

In [9]:
from zutils import Student, Teacher, Doctor, Ward

student = Student("Nam", yob=2000, grade="12th")
print(student.name)
print(student.yob)
print(student.grade)
print(student.describe())

student1 = Student("Hung", yob=2001, grade="11")
print(student1.name)
print(student1.yob)
print(student1.grade)
print(student1.describe())

Nam
2000
12th
Student - Name: Nam - YoB: 2000 - Grade: 12th
Hung
2001
11
Student - Name: Hung - YoB: 2001 - Grade: 11


In [10]:
teacher = Teacher("Hoa", 1999, "Math")
print(teacher.name)
print(teacher.yob)
print(teacher.subject)
teacher.describe()

Hoa
1999
Math


'Teacher - Name: Hoa - YoB: 1999 - Subject: Math'

In [11]:
doctor = Doctor("Smith", 1969, "Dental")
print(doctor.name)
print(doctor.yob)
print(doctor.specialist)
doctor.describe()

Smith
1969
Dental


'Doctor - Name: Smith - YoB: 1969 - Specialist: Dental'

In [12]:
ward = Ward(ward_name="Ben Nghe")
ward.add_people(person=student)
ward.add_people(person=student1)
ward.add_people(person=teacher)
ward.add_people(person=Teacher(name="Linh", yob=2005, subject="Literature"))
ward.add_people(person=doctor)
ward.describe()

Ward Namw: Ben Nghe
Student - Name: Nam - YoB: 2000 - Grade: 12th
Student - Name: Hung - YoB: 2001 - Grade: 11
Teacher - Name: Hoa - YoB: 1999 - Subject: Math
Teacher - Name: Linh - YoB: 2005 - Subject: Literature
Doctor - Name: Smith - YoB: 1969 - Specialist: Dental


In [13]:
# Count Doctor
ward.count_docter()

1

In [14]:
ward.compute_teachers_yob_avg()

2002.0

# 3. Thực hiện xây dựng class Stack với các phương thức (method) sau đây

• initialization method nhận một input "capacity": dùng để khởi tạo stack với capacity là số lượng element mà stack có thể chứa

• .is_empty(): kiểm tra stack có đang rỗng

• .is_full(): kiểm tra stack đã full chưa

• .pop(): loại bỏ top element và trả về giá trị đó

• .push(value) add thêm value vào trong stack

• .top() lấy giá trị top element hiện tại của stack, nhưng không loại bỏ giá trị đó

In [15]:
from zutils import Stack

stack = Stack(capacity=3)

print(stack.is_empty())

stack.push(1)
stack.describe()
print(stack.is_empty())
stack.push(2)
stack.describe()
stack.push(3)
stack.describe()
stack.push(4)
stack.describe()
stack.pop()

True
[1]
False
[1, 2]
[1, 2, 3]
The stack is full.
[1, 2, 3]


3

In [16]:
stack.push(4)
stack.push(5)
stack.describe()
print(stack.pop())
stack.describe()

The stack is full.
[1, 2, 4]
4
[1, 2]


In [17]:
stack.top()

2

# 4. Thực hiện xây dựng class Queue với các chức năng (method) sau đây

• initialization method nhận một input "capacity": dùng để khởi tạo queue với capacity là số lượng element mà queue có thể chứa

• .is_empty(): kiểm tra queue có đang rỗng

• .is_full(): kiểm tra queue đã full chưa

• .dequeue(): loại bỏ first element và trả về giá trị đó

• .enqueue(value) add thêm value vào trong queue

• .front() lấy giá trị first element hiện tại của queue, nhưng không loại bỏ giá trị đó

In [18]:
from zutils import Queue

q = Queue(capacity=4)

# check is empty 
print(q.is_empty())

# check full:
print(q.is_full())

# Add value
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.enqueue(4)

#Print the queue
q.describe()
q.enqueue(5)
# Get first element
q.front()

# remove first element
q.dequeue()
q.describe()




True
False
[1, 2, 3, 4]
The queue is full.
1
1
[2, 3, 4]


# 5. Câu hỏi trắc nghiệm

## 5.1 Kết quả của đoạn code dưới đây là bao nhiêu
import torch


2 import torch .nn as nn

3 4

data = torch . Tensor ([1 , 2, 3])

5 softmax_function = nn. Softmax (dim =0)

6 output = softmax_function ( data )

7 assert round ( output [0]. item () , 2) == 0.09

8 output

a) [0.0900, 0.2747, 0.6652]

b) [0.0700, 0.2447, 0.6652]

c) [0.0900, 0.2447, 0.6652]

d) [0.1900, 0.2447, 0.6692]


In [19]:
data = torch.Tensor ([1 , 2, 3])
softmax(data)

tensor([0.0900, 0.2447, 0.6652])

## 5.2 Hoàn thành đoạn code sau đây theo công thức tính softmax.


import torch
import torch .nn as nn

class MySoftmax (nn. Module ):
def __init__ ( self ):
super (). __init__ ()

def forward (self , x):
======Code here ======

data = torch . Tensor ([5 , 2, 4])
my_softmax = MySoftmax ()
output = my_softmax ( data )
assert round ( output [ -1]. item () , 2) == 0.26
17 output


a) [0.0900, 0.2747, 0.6652]

b) [0.7054, 0.0351, 0.2595]

c) [0.0900, 0.2447, 0.6652]

d) [0.7054, 0.0351, 0.009]

In [20]:
import torch
import torch .nn as nn

class MySoftmax (nn. Module ):
    def __init__ ( self ):
        super (). __init__ ()

    def forward (self , x):
    ### Your Code Here
        exp_x = torch.exp(x)
        exp_sum = torch.sum(exp_x)
        softmax = exp_x / exp_sum 
        return softmax
    ### End Code Here

data = torch . Tensor ([5 , 2, 4])
my_softmax = MySoftmax ()
output = my_softmax ( data )
print(output)
assert round ( output [ -1]. item () , 2) == 0.26

tensor([0.7054, 0.0351, 0.2595])


## 5.3 Hoàn thành đoạn code sau đây theo công thức tính softmax 

a) [0.0900, 0.2747, 0.6652]

b) [0.7054, 0.0351, 0.2595]

c) [0., 0., nan]

d) [0.7054, 0.0351, 0.009]

In [21]:
class MySoftmax (nn.Module):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)

    def forward(self, x):
        exp_x = torch.exp(x)
        exp_sum = torch.sum(exp_x)
        softmax = exp_x / exp_sum 
        return softmax

data = torch . Tensor ([1 , 2, 300000000])
my_softmax = MySoftmax ()
output = my_softmax ( data )

assert round ( output [0]. item () , 2) == 0.0
output

tensor([0., 0., nan])

## 5.4 Hoàn thành đoạn code sau đây theo công thức tính stable softmax
a) [0.0900, 0.2747, 0.6652]

b) [0.0900, 0.2447, 0.6652]

c) [0., 0., nan]

d) [0.7054, 0.0351, 0.009]

In [22]:
data = torch . Tensor ([1 , 2, 3])
softmax_stable = Softmax_stable()
output = softmax_stable ( data )
assert round ( output [ -1]. item () , 2) == 0.67
output

tensor([0.0900, 0.2447, 0.6652])

## 5.5 Một người (person) có thể là student, doctor, hoặc teacher. Một student gồm có name
(string), yob (int) (năm sinh), và grade (string). Các bạn thực hiện viết class Student theo mô tả
trên (Các bạn sẽ viết thêm describe() method để print ra tất cả thông tin của object) và kết quả
đầu ra là gì? Chọn đáp án đúng nhất bên dưới.

a) Student - Name: studentZ2023 - YoB: 2011 - Grade: 6

b) Student - Name: studentZ2023 - YoB: 6 - Grade: 2011

c) Student - Name: 6 - YoB: studentZ2023 - Grade: 2011

d) Tất cả đều sai


In [27]:
student1 = Student ( name =" studentZ2023 ", yob =2011 , grade ="6")
assert student1.yob == 2011
student1.describe ()

'Student - Name:  studentZ2023  - YoB: 2011 - Grade: 6'

## 5.6 Một người (person) có thể là student, doctor, hoặc teacher. Một teacher gồm có name
(string), yob (int), và subject (string). Các bạn thực hiện viết class Teacher theo mô tả trên (Các
bạn sẽ viết thêm describe() method để print ra tất cả thông tin của object) và kết quả đầu ra là
gì? Chọn đáp án đúng nhất bên dưới.

a) Teacher - Name: 1991 - YoB: teacherZ2023 - Subject: History

b) Teacher - Name: teacherZ2023 - YoB: 1991 - Subject: History

c) Teacher - Name: History - YoB: teacherZ2023 - Subject: 1991

d) Tất cả đều sai

In [29]:
from zutils import Teacher
teacher1 = Teacher ( name =" teacherZ2023 ", yob =1991 , subject =" History ")
assert teacher1.yob == 1991
teacher1 . describe ()

'Teacher - Name:  teacherZ2023  - YoB: 1991 - Subject:  History '

## 5.7 Một người (person) có thể là student, doctor, hoặc teacher. Một doctor gồm có name
(string), yob (string), và specialist (string). Các bạn thực hiện viết class Teacher theo mô tả trên
(Các bạn sẽ viết thêm describe() method để print ra tất cả thông tin của object) và kết quả đầu
ra là gì? Chọn đáp án đúng nhất bên dưới.

a) Doctor - Name: doctorZ2023 - YoB: 1981 - Specialist: Endocrinologists

b) Doctor - Name: 1981 - YoB: doctorZ2023 - Specialist: Endocrinologists

c) Teacher - Name: History - YoB: teacherZ2023 - Subject: 1991

d) Tất cả đều sai

In [30]:
from zutils import Doctor

doctor1 = Doctor ( name =" doctorZ2023 ", yob =1981 , specialist =" Endocrinologists ")
assert doctor1.yob == 1981
doctor1 . describe ()

'Doctor - Name:  doctorZ2023  - YoB: 1981 - Specialist:  Endocrinologists '

## 5.8 Một Ward gồm có name (string) và danh sách của mọi người trong Ward. Một người
person có thể là student, doctor, hoặc teacher và cần sử dụng một list để chứa danh sách của mọi
người trong Ward. Viết add_person(person) method trong Ward class để add thêm một người
mới với nghề nghiệp bất kỳ (student, teacher, doctor) vào danh sách người của ward. Tạo ra một
ward object, và thêm vào 1 student, 2 teacher, và 2 doctor. Thực hiện describe() method để in ra
tên ward (name) và toàn bộ thông tin của từng người trong ward. Chọn đáp án đúng nhất bên
dưới cho phương thức đếm số lượng doctor.

a) 4

b) 3

c) 2

d) 1

In [3]:
from zutils import Ward, Student, Teacher, Doctor
student1 = Student ( name =" studentA ", yob =2010 , grade ="7")
teacher1 = Teacher ( name =" teacherA ", yob =1969 , subject =" Math ")
teacher2 = Teacher ( name =" teacherB ", yob =1995 , subject =" History ")
doctor1 = Doctor ( name =" doctorA ", yob =1945 , specialist =" Endocrinologists ")
ward1 = Ward("Ward1")
# assert ward1 . count_doctor () == 1
doctor2 = Doctor ( name =" doctorB ", yob =1975 , specialist =" Cardiologists ")

ward1 . add_people ( student1 )
ward1 . add_people ( teacher1 )
ward1 . add_people ( teacher2 )
ward1 . add_people ( doctor1 )
ward1 . add_people ( doctor2 )
ward1 . count_doctor ()

2

## 5.9 Thực hiện xây dựng class Stack với các chức năng (method) sau đây: initial- ization
method nhận một input "capacity": dùng để khởi tạo stack với capacity là số lượng element mà
stack có thể chứa. .is_full(): kiểm tra stack đã full chưa. .push(value) add thêm value vào trong
stack. Kết quả đầu ra là gì

a) True

b) False

c) None

d) Raise an error

In [4]:
from zutils import Stack

stack1 = Stack ( capacity =5)
stack1 . push (1)
assert stack1 . is_full () == False
stack1 . push (2)
print ( stack1 . is_full ())

False


In [5]:
stack1.describe()

[1, 2]


## 5.10 Thực hiện xây dựng class Stack với các chức năng (method) sau đây: initialization
method nhận một input "capacity", dùng để khởi tạo stack với capacity là số lượng element mà
stack có thể chứa. Phương thức is_empty(): kiểm tra stack có đang rỗng. Phương thức is_full():
kiểm tra stack đã full chưa. Phương thức push(value) add thêm value vào trong stack. Phương
thức top() lấy giá trị top element hiện tại của stack, nhưng không loại bỏ giá trị đó. Kết quả đầu
ra là gì?


a) 1

b) 2

c) None

d) Raise an error

In [8]:
stack1 = Stack ( capacity =5)
stack1 . push (1)
assert stack1 . is_full () == False
stack1 . push (2)
print ( stack1 . top ())

2


## 5.2 Thực hiện xây dựng class Queue với các chức năng (method) sau đây: initialization
method nhận một input "capacity", dùng để khởi tạo queue với capacity là số lượng element mà
queue có thể chứa. Phương thức is_full(): kiểm tra queue đã full chưa. Phương thức enqueue(value)
add thêm value vào trong queue. Kết quả đầu ra là gi?

a) False

b) True

c) None

d) Raise an error

In [6]:
from zutils import Queue
queue1 = Queue ( capacity =5)
queue1 . enqueue (1)
assert queue1 . is_full () == False
queue1 . enqueue (2)
print ( queue1 . is_full ())

False


In [7]:
queue1.describe()

[1, 2]


## 5.12 Thực hiện xây dựng class Queue với các chức năng (method) sau đây: initialization
method nhận một input "capacity" dùng để khởi tạo queue với capacity là số lượng element mà
queue có thể chứa. Phương thức is_full(): kiểm tra queue đã full chưa. Phương thức enqueue(value)
add thêm value vào trong queue. Phương thức front() lấy giá trị first element hiện tại của queue,
nhưng không loại bỏ giá trị đó. Kết quả đầu ra là gì?

a) 4

b) 3

c) 2

d) 1

In [1]:
from zutils import Queue
queue1 = Queue ( capacity =5)
queue1 . enqueue (1)
assert queue1 . is_full () == False
queue1 . enqueue (2)
print ( queue1 . front ())

1


In [2]:
queue1.describe()

[1, 2]
